### PR TITLE
Use dtype utils in providers + add utilities to create `ArrayType` and `EnumType`

### PR DESCRIPTION
**Pull request by axelboc**

### CODE REVIEW

File: packages/app/src/providers/hsds/utils.ts [hunk 1]
```diff
@@ -90,19 +102,25 @@ function convertHsdsCompoundType(
...
-  };
+  );
+}
+
+function convertHsdsEnumType(hsdsType: HsdsEnumType): EnumType | BooleanType {
+  const { base, mapping } = hsdsType;
+  assertHsdsNumericType(base);
+
+  const type = enumType(convertHsdsNumericType(base), mapping);
+  return isBoolEnumType(type) ? boolType() : type; // booleans stored as enums by h5py
```
Comment (axelboc) [inline packages/app/src/providers/hsds/utils.ts:123]: I've also added a utility to abstract away from the providers the check of whether an enum dtype corresponds to the boolean dtype.

File: packages/app/src/providers/hsds/utils.ts [hunk 2]
```diff
@@ -115,37 +133,25 @@ export function convertHsdsType(hsdsType: HsdsType): DType {
...
+      );
 
     case 'H5T_ARRAY':
     case 'H5T_VLEN':
       return {
         class:
           hsdsType.class === 'H5T_ARRAY' ? DTypeClass.Array : DTypeClass.VLen,
         base: convertHsdsType(hsdsType.base),
-        dims: hsdsType.dims,
+        ...(hsdsType.dims && { dims: hsdsType.dims }),
```
Comment (axelboc) [inline packages/app/src/providers/hsds/utils.ts:147]: We don't want `dims: undefined`. This was caught by switching from `toEqual` to `toStrictEqual` in the tests.

File: packages/app/src/providers/mock/mock-file.ts [hunk 3]
```diff
@@ -52,6 +54,26 @@ export function makeMockFile(): GroupWithChildren {
...
+          ],
+        }),
+        scalar('scalar_enum', 2, {
+          type: enumType(intType(8, true), { FOO: 0, BAR: 1, BAZ: 2 }),
+          attributes: [
+            scalarAttr('attr', 2, {
+              type: enumType(intType(8, true), { FOO: 0, BAR: 1, BAZ: 2 }),
+            }),
+          ],
+        }),
```
Comment (axelboc) [inline packages/app/src/providers/mock/mock-file.ts:76]: Adding some mock scalar datasets to test `ArrayType` (with fixed or variable length), and `EnumType`.

File: packages/shared/src/hdf5-utils.ts [hunk 4]
```diff
@@ -33,7 +35,7 @@ export function buildEntityPath(
 /* ----- TYPES ----- */
 
 export function intType(
-  size: 8 | 16 | 32 | 64 = 32,
+  size = 32,
```
Comment (axelboc) [inline packages/shared/src/hdf5-utils.ts:38]: Had to loosen this up to allow creating numeric types from computed sizes - i.e. `length * 8`.

File: packages/app/src/providers/h5grove/utils.test.ts [hunk 5]
```diff
@@ -1,100 +1,69 @@
...
-      size: 32,
-      endianness: Endianness.LE,
-    });
-    expect(convertH5GroveDtype('>u8')).toEqual({
-      class: DTypeClass.Unsigned,
-      size: 64,
-      endianness: Endianness.BE,
-    });
+    expect(convertH5GroveDtype('<i4')).toStrictEqual(
+      intType(32, false, Endianness.LE),
```
Comment (loichuder) [inline packages/app/src/providers/h5grove/utils.test.ts:17]: Not a huge fan of the fact we have to inspect `intType` to know what this `false/true` is for.

Add a helper `uintType` instead and keep `intType` for signed integers ?

File: packages/app/src/providers/hsds/utils.ts (hunk 2 above)
Comment (loichuder) [reply packages/app/src/providers/hsds/utils.ts:147]: :fishing_pole_and_fish:

**Review by loichuder — APPROVED** [review]

File: packages/app/src/providers/h5grove/utils.test.ts (hunk 5 above)
Comment (axelboc) [reply packages/app/src/providers/h5grove/utils.test.ts:17]: I hesitated but you're right, let's do that.

**Comment by axelboc** [comment]
- `intType` vs `uintType`: https://github.com/silx-kit/h5web/pull/1542/commits/4b7273115ce6257866941ada8d966c5ce2cc12f7
- I've also pushed a commit to strengthen `CompoundType` https://github.com/silx-kit/h5web/pull/1542/commits/efb0d541946da8df471244fc5377f259ec81c0c8

File: packages/shared/src/hdf5-utils.ts [hunk 6]
```diff
@@ -73,14 +67,41 @@ export function cplxType(
...
   return { class: DTypeClass.Compound, fields };
 }
 
-export function printableCompoundType(
-  fields: Record<string, PrintableType>,
-): PrintableCompoundType {
-  return { class: DTypeClass.Compound, fields };
+export const printableCompoundType = compoundType<
+  Record<string, PrintableType>
+>;
```
Comment (axelboc) [inline packages/shared/src/hdf5-utils.ts:78]: Taking advantage of a quite recent TypeScript syntax here: https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#instantiation-expressions

**Review by loichuder — APPROVED** [review]

**Pull request merged**